src/objective_types.jl

```
@@ -1,7 +1,11 @@
-function OnceDifferentiable(f, x_seed::AbstractArray{T}, F = zero(T); autodiff = :finite) where T
+function OnceDifferentiable(f, x_seed::AbstractArray{T}, F::Real = real(zero(T)),
+                            DF::AbstractArray = NLSolversBase.alloc_DF(x_seed, F);
+                            autodiff = :finite) where T
     if autodiff == :finite
+        # TODO: Allow user to specify Val{:central}, Val{:forward}, :Val{:complex} (requires care when using :forward I think)
```

@dextorious might be able to correct me, but I think the vast majority of users are better off with central differences, so in my mind it is OK that we only expose that. Complex should simply depend on `T` right?

---

The TODO was there just in case someone wants to fix it in the future :p I would not bother using :central or :complex here myself.

`Val{:complex}` here means calculating finite differences of a function `f:R -> R` by taking a step in a complex direction. See here https://blogs.mathworks.com/cleve/2013/10/14/complex-step-differentiation/

---

ah, yes. I remember looking that up at some point when looking through DiffEqDiffTools. Yeah just leave it there.

---

> Val{:complex} here means calculating finite differences of a function f:R -> R by taking a step in a complex direction. See here https://blogs.mathworks.com/cleve/2013/10/14/complex-step-differentiation/

That's a very cute trick! Probably better to rename it complex_step or something so that it is clear it refers to a specific algorithm and is not intended for complex functions. (also this kind of functionality is dangerous because it silently fails when the function is not given in complex analytic form, even if it is real analytic and perfectly good for differencing, e.g. x -> abs(x)^2 - which sounds ridiculous as written but is perhaps less when written norm(x)^2 or something)

---

The naming isn't ideal, though in the literature it's almost uniformly referred to as "complex-step differentiation". As the API becomes more stabilized and we move on to finally writing some documentation for all this, this and other rough edges will probably be improved along the way. It's true that it has limited applicability and I wouldn't suggest using it as a default recommendation for anyone, but it does give extremely good accuracy quite cheaply, which is sometimes useful.

As for complex vs forward, I'd say pretty much the only reason to ever use forward differences is that they're substantially faster (on the order of 2x usually). If accuracy isn't an issue and for some reason the differentiation is very performance sensitive, I'd recommend `Val{:forward}`. Central differences are the sensible default.